{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1e8098c4",
   "metadata": {},
   "source": [
    "# Examining GangSTR STR genotypes based on bowtie2 alignments\n",
    "### BIO392 05.10.2022\n",
    "\n",
    "Contact: Max Verbiest (maxadriaan.verbiest@uzh.ch)\n",
    "\n",
    "In this notebook, we will explore whether the aligments we generated using bowtie2 are suitable to genotype STR lengths with."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a79d1f24",
   "metadata": {},
   "source": [
    "### 1: Load libraries and files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "b0cb541a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "import gangstr_utils\n",
    "\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "368a9a6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_str_loci = pd.read_csv(\n",
    "    \"../data/repeats/APC_repeats.tsv\", \n",
    "    sep=\"\\t\", \n",
    "    header=None,\n",
    "    names=[\"chr\", \"start\", \"end\", \"unit_len\", \"unit\"])\n",
    "\n",
    "df_gangstr_mut = gangstr_utils.load_gangstr_output(df_str_loci, \"../results/mutated/mutated.vcf\")\n",
    "df_gangstr_wt = gangstr_utils.load_gangstr_output(df_str_loci, \"../results/wildtype/wildtype.vcf\")\n",
    "\n",
    "df_gangstr_merged = (gangstr_utils\n",
    "                         .merge_samples(df_gangstr_mut, df_gangstr_wt)\n",
    "                         .rename(columns={\"alt_s1\": \"alt_mut\", \"alt_s2\": \"alt_wt\"}))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7cc65148",
   "metadata": {},
   "source": [
    "### 2: Look for loci where the samples differ in STR copy number"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e2442531",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>unit_len</th>\n",
       "      <th>unit</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt_mut</th>\n",
       "      <th>alt_wt</th>\n",
       "      <th>sample_difference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr5</td>\n",
       "      <td>21005</td>\n",
       "      <td>21014</td>\n",
       "      <td>2</td>\n",
       "      <td>AT</td>\n",
       "      <td>5.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>chr5</td>\n",
       "      <td>117685</td>\n",
       "      <td>117718</td>\n",
       "      <td>2</td>\n",
       "      <td>AC</td>\n",
       "      <td>17.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>chr5</td>\n",
       "      <td>137481</td>\n",
       "      <td>137490</td>\n",
       "      <td>2</td>\n",
       "      <td>AG</td>\n",
       "      <td>5.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     chr   start     end  unit_len unit   ref  alt_mut  alt_wt  \\\n",
       "12  chr5   21005   21014         2   AT   5.0      2.0     5.0   \n",
       "68  chr5  117685  117718         2   AC  17.0     15.0    14.0   \n",
       "86  chr5  137481  137490         2   AG   5.0      NaN     2.0   \n",
       "\n",
       "    sample_difference  \n",
       "12                3.0  \n",
       "68               -1.0  \n",
       "86                NaN  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gangstr_merged[df_gangstr_merged[\"sample_difference\"] != 0]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0996b008",
   "metadata": {},
   "source": [
    "Strange! From the analyses we performed last week, we know that there should only be a difference in STR length between these two samples at one locus (the AG repeat at chr5:137481-137490). We now see differences at two other loci too. Furthermore, we were not even able to get an accurate estimate of STR length for chr5:137481-137490 in the mutated sample (you can look in the GangSTR vcf file for this sample to see what's going on). \n",
    "\n",
    "At every STR locus except for chr5:137481-137490, the samples should match the reference sequence (because I generated the inputs reads we used directly from the reference sequence). However, if we check the dataframes for STRs where the sample length does not match the reference length, we get many results:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "30a84875",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>unit_len</th>\n",
       "      <th>unit</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr5</td>\n",
       "      <td>9390</td>\n",
       "      <td>9399</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>10.0</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>chr5</td>\n",
       "      <td>32477</td>\n",
       "      <td>32485</td>\n",
       "      <td>1</td>\n",
       "      <td>T</td>\n",
       "      <td>9.0</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>chr5</td>\n",
       "      <td>41515</td>\n",
       "      <td>41523</td>\n",
       "      <td>1</td>\n",
       "      <td>T</td>\n",
       "      <td>9.0</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>chr5</td>\n",
       "      <td>47414</td>\n",
       "      <td>47426</td>\n",
       "      <td>1</td>\n",
       "      <td>T</td>\n",
       "      <td>13.0</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>chr5</td>\n",
       "      <td>58315</td>\n",
       "      <td>58328</td>\n",
       "      <td>1</td>\n",
       "      <td>T</td>\n",
       "      <td>14.0</td>\n",
       "      <td>11.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>chr5</td>\n",
       "      <td>61748</td>\n",
       "      <td>61765</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>18.0</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>chr5</td>\n",
       "      <td>70019</td>\n",
       "      <td>70036</td>\n",
       "      <td>1</td>\n",
       "      <td>T</td>\n",
       "      <td>18.0</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>chr5</td>\n",
       "      <td>74772</td>\n",
       "      <td>74785</td>\n",
       "      <td>2</td>\n",
       "      <td>AC</td>\n",
       "      <td>7.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>chr5</td>\n",
       "      <td>76053</td>\n",
       "      <td>76066</td>\n",
       "      <td>1</td>\n",
       "      <td>T</td>\n",
       "      <td>14.0</td>\n",
       "      <td>13.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>chr5</td>\n",
       "      <td>85497</td>\n",
       "      <td>85508</td>\n",
       "      <td>4</td>\n",
       "      <td>CATA</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>chr5</td>\n",
       "      <td>104055</td>\n",
       "      <td>104098</td>\n",
       "      <td>4</td>\n",
       "      <td>TTTA</td>\n",
       "      <td>11.0</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>chr5</td>\n",
       "      <td>104639</td>\n",
       "      <td>104654</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>16.0</td>\n",
       "      <td>13.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>chr5</td>\n",
       "      <td>107100</td>\n",
       "      <td>107117</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>18.0</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>chr5</td>\n",
       "      <td>117685</td>\n",
       "      <td>117718</td>\n",
       "      <td>2</td>\n",
       "      <td>AC</td>\n",
       "      <td>17.0</td>\n",
       "      <td>14.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>chr5</td>\n",
       "      <td>118533</td>\n",
       "      <td>118552</td>\n",
       "      <td>1</td>\n",
       "      <td>T</td>\n",
       "      <td>20.0</td>\n",
       "      <td>17.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>chr5</td>\n",
       "      <td>124089</td>\n",
       "      <td>124103</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>15.0</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>chr5</td>\n",
       "      <td>124240</td>\n",
       "      <td>124254</td>\n",
       "      <td>3</td>\n",
       "      <td>TAA</td>\n",
       "      <td>5.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>chr5</td>\n",
       "      <td>125960</td>\n",
       "      <td>125976</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>17.0</td>\n",
       "      <td>16.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>chr5</td>\n",
       "      <td>137481</td>\n",
       "      <td>137490</td>\n",
       "      <td>2</td>\n",
       "      <td>AG</td>\n",
       "      <td>5.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     chr   start     end  unit_len  unit   ref   alt\n",
       "2   chr5    9390    9399         1     A  10.0   7.0\n",
       "18  chr5   32477   32485         1     T   9.0   6.0\n",
       "21  chr5   41515   41523         1     T   9.0   6.0\n",
       "23  chr5   47414   47426         1     T  13.0  12.0\n",
       "30  chr5   58315   58328         1     T  14.0  11.0\n",
       "34  chr5   61748   61765         1     A  18.0  15.0\n",
       "39  chr5   70019   70036         1     T  18.0  15.0\n",
       "42  chr5   74772   74785         2    AC   7.0   4.0\n",
       "45  chr5   76053   76066         1     T  14.0  13.0\n",
       "50  chr5   85497   85508         4  CATA   3.0   1.0\n",
       "60  chr5  104055  104098         4  TTTA  11.0   8.0\n",
       "61  chr5  104639  104654         1     A  16.0  13.0\n",
       "63  chr5  107100  107117         1     A  18.0  15.0\n",
       "68  chr5  117685  117718         2    AC  17.0  14.0\n",
       "69  chr5  118533  118552         1     T  20.0  17.0\n",
       "72  chr5  124089  124103         1     A  15.0  12.0\n",
       "74  chr5  124240  124254         3   TAA   5.0   2.0\n",
       "76  chr5  125960  125976         1     A  17.0  16.0\n",
       "86  chr5  137481  137490         2    AG   5.0   2.0"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gangstr_wt[df_gangstr_wt[\"ref\"] != df_gangstr_wt[\"alt\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "86b9b2f2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>unit_len</th>\n",
       "      <th>unit</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr5</td>\n",
       "      <td>9390</td>\n",
       "      <td>9399</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>10.0</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr5</td>\n",
       "      <td>21005</td>\n",
       "      <td>21014</td>\n",
       "      <td>2</td>\n",
       "      <td>AT</td>\n",
       "      <td>5.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>chr5</td>\n",
       "      <td>32477</td>\n",
       "      <td>32485</td>\n",
       "      <td>1</td>\n",
       "      <td>T</td>\n",
       "      <td>9.0</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>chr5</td>\n",
       "      <td>41515</td>\n",
       "      <td>41523</td>\n",
       "      <td>1</td>\n",
       "      <td>T</td>\n",
       "      <td>9.0</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>chr5</td>\n",
       "      <td>47414</td>\n",
       "      <td>47426</td>\n",
       "      <td>1</td>\n",
       "      <td>T</td>\n",
       "      <td>13.0</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>chr5</td>\n",
       "      <td>58315</td>\n",
       "      <td>58328</td>\n",
       "      <td>1</td>\n",
       "      <td>T</td>\n",
       "      <td>14.0</td>\n",
       "      <td>11.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>chr5</td>\n",
       "      <td>61748</td>\n",
       "      <td>61765</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>18.0</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>chr5</td>\n",
       "      <td>70019</td>\n",
       "      <td>70036</td>\n",
       "      <td>1</td>\n",
       "      <td>T</td>\n",
       "      <td>18.0</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>chr5</td>\n",
       "      <td>74772</td>\n",
       "      <td>74785</td>\n",
       "      <td>2</td>\n",
       "      <td>AC</td>\n",
       "      <td>7.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>chr5</td>\n",
       "      <td>76053</td>\n",
       "      <td>76066</td>\n",
       "      <td>1</td>\n",
       "      <td>T</td>\n",
       "      <td>14.0</td>\n",
       "      <td>13.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>chr5</td>\n",
       "      <td>85497</td>\n",
       "      <td>85508</td>\n",
       "      <td>4</td>\n",
       "      <td>CATA</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>chr5</td>\n",
       "      <td>104055</td>\n",
       "      <td>104098</td>\n",
       "      <td>4</td>\n",
       "      <td>TTTA</td>\n",
       "      <td>11.0</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>chr5</td>\n",
       "      <td>104639</td>\n",
       "      <td>104654</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>16.0</td>\n",
       "      <td>13.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>chr5</td>\n",
       "      <td>107100</td>\n",
       "      <td>107117</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>18.0</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>chr5</td>\n",
       "      <td>117685</td>\n",
       "      <td>117718</td>\n",
       "      <td>2</td>\n",
       "      <td>AC</td>\n",
       "      <td>17.0</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>chr5</td>\n",
       "      <td>118533</td>\n",
       "      <td>118552</td>\n",
       "      <td>1</td>\n",
       "      <td>T</td>\n",
       "      <td>20.0</td>\n",
       "      <td>17.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>chr5</td>\n",
       "      <td>124089</td>\n",
       "      <td>124103</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>15.0</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>chr5</td>\n",
       "      <td>124240</td>\n",
       "      <td>124254</td>\n",
       "      <td>3</td>\n",
       "      <td>TAA</td>\n",
       "      <td>5.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>chr5</td>\n",
       "      <td>125960</td>\n",
       "      <td>125976</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>17.0</td>\n",
       "      <td>16.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>chr5</td>\n",
       "      <td>137481</td>\n",
       "      <td>137490</td>\n",
       "      <td>2</td>\n",
       "      <td>AG</td>\n",
       "      <td>5.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     chr   start     end  unit_len  unit   ref   alt\n",
       "2   chr5    9390    9399         1     A  10.0   7.0\n",
       "12  chr5   21005   21014         2    AT   5.0   2.0\n",
       "18  chr5   32477   32485         1     T   9.0   6.0\n",
       "21  chr5   41515   41523         1     T   9.0   6.0\n",
       "23  chr5   47414   47426         1     T  13.0  12.0\n",
       "30  chr5   58315   58328         1     T  14.0  11.0\n",
       "34  chr5   61748   61765         1     A  18.0  15.0\n",
       "39  chr5   70019   70036         1     T  18.0  15.0\n",
       "42  chr5   74772   74785         2    AC   7.0   4.0\n",
       "45  chr5   76053   76066         1     T  14.0  13.0\n",
       "50  chr5   85497   85508         4  CATA   3.0   1.0\n",
       "60  chr5  104055  104098         4  TTTA  11.0   8.0\n",
       "61  chr5  104639  104654         1     A  16.0  13.0\n",
       "63  chr5  107100  107117         1     A  18.0  15.0\n",
       "68  chr5  117685  117718         2    AC  17.0  15.0\n",
       "69  chr5  118533  118552         1     T  20.0  17.0\n",
       "72  chr5  124089  124103         1     A  15.0  12.0\n",
       "74  chr5  124240  124254         3   TAA   5.0   2.0\n",
       "76  chr5  125960  125976         1     A  17.0  16.0\n",
       "86  chr5  137481  137490         2    AG   5.0   NaN"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gangstr_mut[df_gangstr_mut[\"ref\"] != df_gangstr_mut[\"alt\"]]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "61cac8e2",
   "metadata": {},
   "source": [
    "### 3: What is going on??\n",
    "\n",
    "So, obviously, something is not quite right. Your task is now to find out where the observed errors come from, and how to fix it. The only tip you'll get for now is that there are problematic alignments in the BAM files we used, and that we want to extract only the 'properly paired' alignments. Luckily, the [samtools](https://www.htslib.org/doc/samtools.html) suite of tools (included in the BIO392_STRs conda env) offers all the functionality needed to address the problem. Particularly [samtools flagstat](https://www.htslib.org/doc/samtools-flagstat.html), [samtools view](https://www.htslib.org/doc/samtools-view.html) and [samtools flags](https://www.htslib.org/doc/samtools-flags.html) will be very useful. Going through these manual pages would be a good start. You can also visually inspect the aligments with [samtools tview](https://www.htslib.org/doc/samtools-tview.html).\n",
    "\n",
    "\n",
    "**Task 1:** Find a way to identify problematic alignments in the BAM files generated by bowtie2. Extract only 'proper pairs', and save these in new BAM files.\n",
    "\n",
    "**Task 2:** Rerun GangSTR using these new BAM files (you will need to change some things in the GangSTR bash script), and inspect the results using the code blocks at the start of this notebook.\n",
    "\n",
    "**Task 3:** Were you able to solve/improve the outcome? Is bowtie2 a good aligner to use for genotyping STRs?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "9b3c77ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Task 1\n",
    "\n",
    "# mutated: 23944 + 0 properly paired (49.97% : N/A)\n",
    "# wildtype: 23980 + 0 properly paired (50.05% : N/A)\n",
    "\n",
    "# samtools view -bf 0x2 /Volumes/UsersBio/mgull/BIO392/day10/2022-10-05/data/alignments/wildtype/APC_mut.bam  > /Volumes/UsersBio/mgull/BIO392/day10/2022-10-05/data/alignments/wildtype/APC_mutt.bam\n",
    "# samtools view -bf 0x2 /Volumes/UsersBio/mgull/BIO392/day10/2022-10-05/data/alignments/wildtype/APC_wt.bam  > /Volumes/UsersBio/mgull/BIO392/day10/2022-10-05/data/alignments/wildtype/APC_wtt.bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "71bfa5d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Task 2\n",
    "\n",
    "df_str_loci = pd.read_csv(\n",
    "    \"../data/repeats/APC_repeats.tsv\", \n",
    "    sep=\"\\t\", \n",
    "    header=None,\n",
    "    names=[\"chr\", \"start\", \"end\", \"unit_len\", \"unit\"])\n",
    "\n",
    "df_gangstr_mut = gangstr_utils.load_gangstr_output(df_str_loci, \"../results/mutated/mutated.vcf\")\n",
    "df_gangstr_wt = gangstr_utils.load_gangstr_output(df_str_loci, \"../results/wildtype/wildtype.vcf\")\n",
    "\n",
    "df_gangstr_merged = (gangstr_utils\n",
    "                         .merge_samples(df_gangstr_mut, df_gangstr_wt)\n",
    "                         .rename(columns={\"alt_s1\": \"alt_mut\", \"alt_s2\": \"alt_wt\"}))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "aa8b2d1c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>unit_len</th>\n",
       "      <th>unit</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt_mut</th>\n",
       "      <th>alt_wt</th>\n",
       "      <th>sample_difference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>chr5</td>\n",
       "      <td>137481</td>\n",
       "      <td>137490</td>\n",
       "      <td>2</td>\n",
       "      <td>AG</td>\n",
       "      <td>5.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     chr   start     end  unit_len unit  ref  alt_mut  alt_wt  \\\n",
       "86  chr5  137481  137490         2   AG  5.0      NaN     5.0   \n",
       "\n",
       "    sample_difference  \n",
       "86                NaN  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gangstr_merged[df_gangstr_merged[\"sample_difference\"] != 0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "cea68a0d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>unit_len</th>\n",
       "      <th>unit</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [chr, start, end, unit_len, unit, ref, alt]\n",
       "Index: []"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gangstr_wt[df_gangstr_wt[\"ref\"] != df_gangstr_wt[\"alt\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "d0ac42ef",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>unit_len</th>\n",
       "      <th>unit</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>chr5</td>\n",
       "      <td>137481</td>\n",
       "      <td>137490</td>\n",
       "      <td>2</td>\n",
       "      <td>AG</td>\n",
       "      <td>5.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     chr   start     end  unit_len unit  ref  alt\n",
       "86  chr5  137481  137490         2   AG  5.0  NaN"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gangstr_mut[df_gangstr_mut[\"ref\"] != df_gangstr_mut[\"alt\"]]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4fa83600",
   "metadata": {},
   "source": [
    "### Task 3    \n",
    "Were you able to solve/improve the outcome?    \n",
    "--> Yes.    <br>       \n",
    "Is bowtie2 a good aligner to use for genotyping STRs?\n",
    "* Bowtie 2 is an ultrafast and memory-efficient tool for aligning sequencing reads to long reference sequences. \n",
    "* It is particularly good at aligning short reads to relatively long genomes. \n",
    "* Bowtie 2 indexes the genome with an FM Index to keep its memory footprint small: for the human genome, its memory footprint is typically around 3.2 gigabytes of RAM. Bowtie 2 supports gapped, local, and paired-end alignment modes. \n",
    "* Multiple processors can be used simultaneously to achieve greater alignment speed.\n",
    "* Bowtie 2 outputs alignments in SAM format, enabling interoperation with a large number of other tools that use SAM. \n",
    "* Bowtie 2 is often the first step in pipelines for comparative genomics.    \n",
    "\n",
    "--> therefore it seems like a good aligner to me."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
