# Sequence analysis and file formats in bioinformatics
## Task 1
* In the ‘scripts’ directory, you will find 5 numbered bash scripts:   
<img width="723" alt="Screen Shot 2022-10-05 at 09 20 56" src="https://user-images.githubusercontent.com/114056296/194003220-657e530e-85e4-49f6-b945-bbbdcd24e85b.png">
    
* For each of the scricpts, write a brief overview of what tasks the script performs and which tools and file formats are used, then run it ('bash <script_name')   
    * 01_unzip_reads.sh: The fastq files are compressed to save space, the following commands will unzip them
    * 02_bowtie2_index.sh: Before we can align the reads to the reference with bowtie2, we need to build bowtie2 indices for reference sequence.
    * 03_alignment_bowtie2.sh: Add read group information to the generated alignment (needed by GangSTR later on)
    * 04_process_alignment.sh: Using samtools view, we convert the alignments generated in the previous step to (binary) BAM format. Then, with samtools sort, we sort the alignments by their starting coordinates. Finally, we generate indices for our alignments with samtools index
    * 05_run_gangstr.sh: We run the GangSTR STR genotyping tool on both our alignments (wildtype and  mutated). GangSTR also needs the reference genome and a tab-separated file indicating where STR loci are located in the reference genome.    

    
## Task 2
* Find a way to identify problematic alignments in the BAM files generated by bowtie2. Extract only 'proper pairs', and save these in new BAM files.
    * mutated: 23944 + 0 properly paired (49.97% : N/A)
    * wildtype: 23980 + 0 properly paired (50.05% : N/A)
    * samtools view -bf 0x2 APC_mut.bam  > APC_mutt.bam
    * samtools view -bf 0x2 /APC_wt.bam  > APC_wtt.bam 
* Go through the ‘check_GangSTR_output.ipynb’ notebook. Make sure to upload the notebook to the github repo once you’re done!
* Were you able to solve/improve the outcome?    
    * Yes.   
    <img width="538" alt="Screen Shot 2022-10-05 at 11 08 54" src="https://user-images.githubusercontent.com/114056296/194024218-46ca4d3d-f9dc-41d5-a590-879f2bc17661.png"> 
       
* Is bowtie2 a good aligner to use for genotyping STRs?
    * Bowtie 2 is an ultrafast and memory-efficient tool for aligning sequencing reads to long reference sequences. 
    * It is particularly good at aligning short reads to relatively long genomes. 
    * Bowtie 2 indexes the genome with an FM Index to keep its memory footprint small: for the human genome, its memory footprint is typically around 3.2 gigabytes of RAM. Bowtie 2 supports gapped, local, and paired-end alignment modes. 
    * Multiple processors can be used simultaneously to achieve greater alignment speed.
    * Bowtie 2 outputs alignments in SAM format, enabling interoperation with a large number of other tools that use SAM. 
    * Bowtie 2 is often the first step in pipelines for comparative genomics.    
    --> Therefore, it seems like a good aligner to me.
